File: src/test/java/org/hisp/dhis/rules/functions/RuleFunctionCeilTests.java
package org.hisp.dhis.rules.functions;

import org.hamcrest.MatcherAssert;
import org.hisp.dhis.parser.expression.CommonExpressionVisitor;
import org.hisp.dhis.parser.expression.antlr.ExpressionParser;
import org.hisp.dhis.parser.expression.function.ScalarFunction;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.runners.MockitoJUnitRunner;

import static org.hamcrest.CoreMatchers.is;
import static org.mockito.Mockito.when;

@RunWith( MockitoJUnitRunner.class )
public class RuleFunctionCeilTests
{

        @Mock
        private ExpressionParser.ExprContext context;

        @Mock
        private CommonExpressionVisitor visitor;

        @Mock
        private ExpressionParser.ExprContext mockedFirstExpr;

        @Before
        public void setUp() {
                when(context.expr(0)).thenReturn( mockedFirstExpr );
        }

        @Test
        public void evaluateMustReturnCeiledValue()
        {
                RuleFunctionCeil ceil = new RuleFunctionCeil();

                when( visitor.castStringVisit( mockedFirstExpr ) ).thenReturn( "4.1" );
                MatcherAssert.assertThat(ceil.evaluate(context, visitor), is("5"));

                when( visitor.castStringVisit( mockedFirstExpr ) ).thenReturn( "0.8" );
                MatcherAssert.assertThat(ceil.evaluate(context, visitor), is("1"));

                when( visitor.castStringVisit( mockedFirstExpr ) ).thenReturn( "5.1" );
                MatcherAssert.assertThat(ceil.evaluate(context, visitor), is("6"));

                when( visitor.castStringVisit( mockedFirstExpr ) ).thenReturn( "1" );
                MatcherAssert.assertThat(ceil.evaluate(context, visitor), is("1"));

                when( visitor.castStringVisit( mockedFirstExpr ) ).thenReturn( "-9.3" );
                MatcherAssert.assertThat(ceil.evaluate(context, visitor), is("-9"));

                when( visitor.castStringVisit( mockedFirstExpr ) ).thenReturn( "-5.9" );
                MatcherAssert.assertThat(ceil.evaluate(context, visitor), is("-5"));
        }

        @Test
        public void return_zero_when_number_is_invalid()
        {
                when( visitor.castStringVisit( mockedFirstExpr ) ).thenReturn( "not a number" );

                ScalarFunction ceil = new RuleFunctionCeil();

                MatcherAssert.assertThat( ceil.evaluate( context, visitor ), is("0") );
        }
}

File: src/jvmTest/kotlin/org/hisp/dhis/rules/functions/RuleFunctionCeilTests.kt
package org.hisp.dhis.rules.functions

import org.hamcrest.CoreMatchers.`is`
import org.hamcrest.MatcherAssert.assertThat
import org.hisp.dhis.rules.RuleVariableValue
import org.junit.Test
import org.junit.runner.RunWith
import org.junit.runners.JUnit4
import java.lang.NullPointerException
import kotlin.test.assertFailsWith

@RunWith(JUnit4::class)
class RuleFunctionCeilTests {

    private val variableValues = HashMap<String, RuleVariableValue>()

    @Test
    fun evaluateMustReturnCeilValue() {
        val ceil = RuleFunctionCeil.create()

        assertThat(ceil.evaluate(listOf("0"), variableValues, null), `is`("0"))
        assertThat(ceil.evaluate(listOf("0.8"), variableValues, null), `is`("1"))
        assertThat(ceil.evaluate(listOf("5.1"), variableValues, null), `is`("6"))
        assertThat(ceil.evaluate(listOf("1"), variableValues, null), `is`("1"))
        assertThat(ceil.evaluate(listOf("-9.3"), variableValues, null), `is`("-9"))
        assertThat(ceil.evaluate(listOf("-5.9"), variableValues, null), `is`("-5"))
    }

    @Test
    fun throw_illegal_argument_exception_when_argument_count_is_greater_than_expected() {
        assertFailsWith<IllegalArgumentException> {
            RuleFunctionCeil.create().evaluate(listOf("5.9", "6.8"), variableValues, null)
        }
    }

    @Test
    fun throw_illegal_argument_exception_when_arguments_count_is_lower_than_expected() {
        assertFailsWith<IllegalArgumentException> {
            RuleFunctionCeil.create().evaluate(emptyList(), variableValues, null)
        }
    }

    @Test
    fun throw_null_pointer_exception_when_arguments_is_null() {
        assertFailsWith<NullPointerException> {
            RuleFunctionCeil.create().evaluate(null!!, variableValues, null)
        }
    }
}
